code/modules/GOAP/GOAPAgent.dm

```diff
@@ -0,0 +1,243 @@
+GLOBAL_LIST_INIT(dangerous_turfs, typecacheof(list(
+	/turf/open/lava,
+	/turf/open/chasm)))
+
+#define MAX_TRIES 10
+#define STATE_IDLE		0
+#define STATE_MOVINGTO	1
+#define STATE_ACTING	2
+
+#define MOVETYPE_ASTAR 1
+#define MOVETYPE_FAKESTAR 2
+#define MOVETYPE_DUMB 3
+
+// #define GOAP_DEBUG_ENABLE "fart"
+
+#ifdef GOAP_DEBUG_ENABLE
+#define goap_debug(string) world.log << "GOAP: [string]"
+#else
+#define goap_debug(string) ;
+#endif
+/datum/goap_agent
```

Comment:
All of this could be made into a component so it can be put onto anything without needing to define a new variable then setting it to a GOAP AGENT 
see russian implementation
`var/datum/goap_agent/goap_ai`

`var/ai_type = /datum/goap_agent/russian_melee`

Reply:
It might not be worth saving the two vars, i'll hold off on converting it just yet

Reply:
This will likely be required to become a component as per codebase standards but I'm not an expert, consult a maintainer.

Reply:
I’ve talked with ninja about it and he said he wouldn’t force that amount of work and that if GOAP was being designed from scratch it’d have been a different story